### ISSUE TITLE
Runtime Crash need help.

### ISSUE THREAD

**Issue by doncat99** [JoshuaC215/agent-service-toolkit#88]
```
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/starlette/responses.py", line 265, in __call__
    await wrap(partial(self.listen_for_disconnect, receive))
  File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/starlette/responses.py", line 261, in wrap
    await func()
  File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/starlette/responses.py", line 238, in listen_for_disconnect
    message = await receive()
  File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/uvicorn/protocols/http/httptools_impl.py", line 596, in receive
    await self.message_event.wait()
  File "/Users/huangdon/miniconda3/envs/ai_writer/lib/python3.10/asyncio/locks.py", line 214, in wait
    await fut
asyncio.exceptions.CancelledError: Cancelled by cancel scope 10faefac0

During handling of the above exception, another exception occurred:

  + Exception Group Traceback (most recent call last):
  |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/uvicorn/protocols/http/httptools_impl.py", line 435, in run_asgi
  |     result = await app(  # type: ignore[func-returns-value]
  |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 78, in __call__
  |     return await self.app(scope, receive, send)
  |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/fastapi/applications.py", line 1054, in __call__
  |     await super().__call__(scope, receive, send)
  |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/starlette/applications.py", line 123, in __call__
  |     await self.middleware_stack(scope, receive, send)
  |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/starlette/middleware/errors.py", line 186, in __call__
  |     raise exc
  |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/starlette/middleware/errors.py", line 164, in __call__
  |     await self.app(scope, receive, _send)
  |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 65, in __call__
  |     await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
  |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
  |     raise exc
  |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
  |     await app(scope, receive, sender)
  |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/starlette/routing.py", line 756, in __call__
  |     await self.middleware_stack(scope, receive, send)
  |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/starlette/routing.py", line 776, in app
  |     await route.handle(scope, receive, send)
  |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/starlette/routing.py", line 297, in handle
  |     await self.app(scope, receive, send)
  |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/starlette/routing.py", line 77, in app
  |     await wrap_app_handling_exceptions(app, request)(scope, receive, send)
  |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
  |     raise exc
  |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
  |     await app(scope, receive, sender)
  |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/starlette/routing.py", line 75, in app
  |     await response(scope, receive, send)
  |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/starlette/responses.py", line 258, in __call__
  |     async with anyio.create_task_group() as task_group:
  |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 763, in __aexit__
  |     raise BaseExceptionGroup(
  | exceptiongroup.ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/starlette/responses.py", line 261, in wrap
    |     await func()
    |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/starlette/responses.py", line 250, in stream_response
    |     async for chunk in self.body_iterator:
    |   File "/Users/huangdon/Documents/AgentCenter/src/service/service.py", line 129, in message_generator
    |     async for event in agent.astream_events(**kwargs, version="v2"):
    |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/langchain_core/runnables/base.py", line 1388, in astream_events
    |     async for event in event_stream:
    |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/langchain_core/tracers/event_stream.py", line 1012, in _astream_events_implementation_v2
    |     await task
    |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/langchain_core/tracers/event_stream.py", line 967, in consume_astream
    |     async for _ in event_streamer.tap_output_aiter(run_id, stream):
    |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/langchain_core/tracers/event_stream.py", line 180, in tap_output_aiter
    |     first = await py_anext(output, default=sentinel)
    |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/langchain_core/utils/aiter.py", line 76, in anext_impl
    |     return await __anext__(iterator)
    |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/langgraph/pregel/__init__.py", line 1511, in astream
    |     async for _ in runner.atick(
    |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/langgraph/pregel/runner.py", line 130, in atick
    |     await arun_with_retry(t, retry_policy, stream=self.use_astream)
    |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/langgraph/pregel/retry.py", line 99, in arun_with_retry
    |     async for _ in task.proc.astream(task.input, config):
    |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/langgraph/utils/runnable.py", line 576, in astream
    |     async for chunk in aiterator:
    |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/langchain_core/tracers/event_stream.py", line 180, in tap_output_aiter
    |     first = await py_anext(output, default=sentinel)
    |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/langchain_core/utils/aiter.py", line 76, in anext_impl
    |     return await __anext__(iterator)
    |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/langchain_core/runnables/base.py", line 1455, in atransform
    |     async for ichunk in input:
    |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/langchain_core/runnables/base.py", line 1018, in astream
    |     yield await self.ainvoke(input, config, **kwargs)
    |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/langgraph/utils/runnable.py", line 238, in ainvoke
    |     ret = await self.afunc(input, **kwargs)
    |   File "/Users/huangdon/Documents/AgentCenter/src/agents/bg_task_agent/bg_task_agent.py", line 41, in bg_task
    |     await task1.start(config=config)
    |   File "/Users/huangdon/Documents/AgentCenter/src/agents/bg_task_agent/task.py", line 31, in start
    |     task_message = await self._generate_and_dispatch_message(config, data)
    |   File "/Users/huangdon/Documents/AgentCenter/src/agents/bg_task_agent/task.py", line 26, in _generate_and_dispatch_message
    |     await task_custom_data.adispatch()
    |   File "/Users/huangdon/Documents/AgentCenter/src/agents/utils.py", line 26, in adispatch
    |     await adispatch_custom_event(
    |   File "/Users/huangdon/Documents/AgentCenter/.venv/lib/python3.10/site-packages/langchain_core/callbacks/manager.py", line 2533, in adispatch_custom_event
    |     raise RuntimeError(msg)
    | RuntimeError: Unable to dispatch an adhoc event without a parent run id.This function can only be called from within an existing run (e.g.,inside a tool or a RunnableLambda or a RunnableGenerator.)If you are doing that and still seeing this error, try explicitlypassing the config parameter to this function.
```
    
Hi, I am facing a crash and need some help to fix it. It doesn't always occur. sometimes I switch from research agent to bg-task agent, it occured.

**Comment by JoshuaC215** [comment]
Thanks - does it only occur when you start a conversation and then switch agent from the Streamlit app? It isn't a fully supported scenario so it might explain the error. We should reset the thread_id and conversation when switching agent in the UI or something like that to resolve.

Is it blocking you somehow or just a crash you observed?

**Comment by doncat99** [comment]
> Thanks - does it only occur when you start a conversation and then switch agent from the Streamlit app? It isn't a fully supported scenario so it might explain the error. We should reset the thread_id and conversation when switching agent in the UI or something like that to resolve.
> 
> Is it blocking you somehow or just a crash you observed?

![errror](https://github.com/user-attachments/assets/a0080c9c-e2db-47df-bfb0-653a9541351e)

It's blocking me, and the conversation dies. I guess it is due to the thread_id too. I am thinking big, much like a conversation may introduce many agents to work together. Thus, somehow is it possible that agents will share history? Or at some points or task milestone, one agent will make a conclusion info scheme for sharing between agents.

**Comment by JoshuaC215** [comment]
No, I'm not planning to support arbitrary agent coordination through this repo.

If the issue is just due to bg_task_agent - it's an experimental feature so not something I'm providing deep support for.

HOWEVER if you want to walk through a full repro of this issue using the original template source code (such as recording a short video or writing out step by step) I can take a closer look. Otherwise I won't be able to help here. Sorry!

**Comment by peterkeppert** [comment]
I managed to reliably reproduce this behavior with Python 3.10. Steps to reproduce:
1. Start service and UI
2. Open new conversation and keep default settings (OpenAI, Research Assistant and Streaming)
3. Ask question
4. AI answers
5. Switch to BG Task Agent
6. Ask next question
7. Error occurs

This happened also if we omitted steps 3-4, that is if first question asked went to BG Task Agent directly.

Using Python 3.12, this scenario worked flawlessly. From the error message, it looked like run ID wasn't propagated correctly and so I identified the root cause being in the kind of behavior described for example here: https://python.langchain.com/docs/how_to/tool_stream_events/
![image](https://github.com/user-attachments/assets/31887603-9eb1-40a4-88aa-0c049def4744)

Fix attached :)
